pkg/controller/meterdefinition/meterdefinition_controller.go

```diff
@@ -142,27 +188,500 @@ func (r *ReconcileMeterDefinition) Reconcile(request reconcile.Request) (reconci
+			ReconcileResult: reconcile.Result{},
+			Err:             err,
+		}
+	}
+
+	return &ExecResult{
+		ReconcileResult: reconcile.Result{Requeue: true},
+		Err:             err,
+	}
+}
+
+func (r *ReconcileMeterDefinition) handleUpdateConflictForStatusCondition(condition status.Condition, request reconcile.Request, reqLogger logr.Logger) *ExecResult {
+	reqLogger.Info("conflict err")
+
+	latestMeterdef := v1alpha1.MeterDefinition{}
+	r.client.Get(context.TODO(), request.NamespacedName, &latestMeterdef)
+
+	latestMeterdef.Status.Conditions.SetCondition(condition)
+	err := r.client.Status().Update(context.TODO(), &latestMeterdef)
+	if err != nil {
+		reqLogger.Error(err, "error updating with resource version port")
+		return &ExecResult{
+			ReconcileResult: reconcile.Result{},
+			Err:             err,
+		}
+	}
+
+	return &ExecResult{
+		ReconcileResult: reconcile.Result{Requeue: true},
+		Err:             err,
+	}
+}
+
+func (r *ReconcileMeterDefinition) handleUpdateConflictForErrorConditions(err error, conditionMsg error, conditionType status.ConditionType, instance *v1alpha1.MeterDefinition, reqLogger logr.Logger, request reconcile.Request) *ExecResult {
+
+	latestMeterdef := v1alpha1.MeterDefinition{}
+	r.client.Get(context.TODO(), request.NamespacedName, &latestMeterdef)
+	reqLogger.Info("conflict err")
+
+	reqLogger.Info("Updating status with error")
+	instance.Status.Conditions.SetCondition(status.Condition{
+		Message: conditionMsg.Error(),
+		Type:    status.ConditionType(conditionType),
+	})
+
+	err = r.client.Status().Update(context.TODO(), &latestMeterdef)
+	if err != nil {
+		reqLogger.Error(err, "error updating with resource version port")
+		return &ExecResult{
+			ReconcileResult: reconcile.Result{},
+			Err:             err,
+		}
+	}
+
+	return &ExecResult{
+		ReconcileResult: reconcile.Result{Requeue: true},
+		Err:             err,
+	}
+}
+
+func (r *ReconcileMeterDefinition) getServiceAccountToken(instance *v1alpha1.MeterDefinition, reqLogger logr.Logger) (string, error) {
+	r.serviceAccountClient.Lock()
+	defer r.serviceAccountClient.Unlock()
+
+	now := metav1.Now().UTC()
+	client,tr := r.newSAClientAndToken(instance)
+	opts := metav1.CreateOptions{}
+
+	if r.serviceAccountClient.Token == nil {
+		reqLogger.Info("auth token from service account found")
+
+		return r.returnToken(client,tr,opts)
+	}
+
+	if now.UTC().After(r.serviceAccountClient.Token.ExpirationTimestamp.Time) {
+
+		reqLogger.Info("service account token is expired")
+
+		return r.returnToken(client,tr,opts)
+	}
+
+	return r.returnToken(client,tr,opts)
+}
+
+func (r *ReconcileMeterDefinition) newSAClientAndToken(instance *v1alpha1.MeterDefinition)(typedv1.ServiceAccountInterface,*authv1.TokenRequest){
+
+	client := r.serviceAccountClient.KubernetesInterface.CoreV1().ServiceAccounts(instance.Namespace)
+
+	tr := &authv1.TokenRequest{
+		Spec: authv1.TokenRequestSpec{
+			Audiences:         []string{"rhm-prometheus-meterbase.openshift-redhat-marketplace.svc"},
```

Comment:
Let's not hardcode this, it'll fail if we switch namespaces.

Reply:
@zach-source changes coming in the library version of this code to make everything configurable